rpcs3/Emu/RSX/overlay_controls.h

```diff
@@ -0,0 +1,1205 @@
+			f32 padding_top = 0.f;
+			f32 padding_bottom = 0.f;
+
+			overlay_element() {}
+			overlay_element(u16 _w, u16 _h) : w(_w), h(_h) {}
+
+			virtual void refresh()
+			{
+				//Just invalidate for draw when get_compiled() is called
+				is_compiled = false;
+			}
+
+			virtual void translate(s16 _x, s16 _y)
+			{
+				x = (u16)(x + _x);
+				y = (u16)(y + _y);
+
+				is_compiled = false;
+			}
+
+			virtual void scale(f32 _x, f32 _y, bool origin_scaling)
+			{
+				if (origin_scaling)
+				{
+					x = (u16)(_x * x);
+					y = (u16)(_y * y);
+				}
+
+				w = (u16)(_x * w);
+				h = (u16)(_y * h);
+
+				is_compiled = false;
+			}
+
+			virtual void set_pos(u16 _x, u16 _y)
+			{
+				x = _x;
+				y = _y;
+
+				is_compiled = false;
+			}
+
+			virtual void set_size(u16 _w, u16 _h)
+			{
+				w = _w;
+				h = _h;
+
+				is_compiled = false;
+			}
+
+			virtual void set_padding(f32 left, f32 right, f32 top, f32 bottom)
+			{
+				padding_left = left;
+				padding_right = right;
+				padding_top = top;
+				padding_bottom = bottom;
+
+				is_compiled = false;
+			}
+
+			virtual void set_padding(f32 padding)
+			{
+				padding_left = padding_right = padding_top = padding_bottom = padding;
+				is_compiled = false;
+			}
+
+			virtual void set_text(std::string& text)
+			{
+				this->text = text;
+				is_compiled = false;
+			}
+
+			virtual void set_text(const char* text)
+			{
+				this->text = text;
+				is_compiled = false;
+			}
+
+			virtual void set_font(const char* font_name, u16 font_size)
+			{
+				font = fontmgr::get(font_name, font_size);
+				is_compiled = false;
+			}
+
+			virtual void align_text(text_align align)
+			{
+				alignment = align;
+				is_compiled = false;
+			}
+
+			virtual std::vector<vertex> render_text(const char *string, f32 x, f32 y)
+			{
+				auto renderer = font;
+				if (!renderer) renderer = fontmgr::get("Arial", 12);
+
+				f32 text_extents_w = 0.f;
+				std::vector<vertex> result = renderer->render_text(string);
+				for (auto &v : result)
+				{
+					//Check for real text region extent
+					//TODO: Ellipsis
+					text_extents_w = std::max(v.values[0], text_extents_w);
+
+					//Apply transform.
+					//(0, 0) has text sitting 50% off the top left corner (text is outside the rect) hence the offset by text height / 2
+					v.values[0] += x + padding_left;
+					v.values[1] += y + padding_top + (f32)renderer->size_px * 0.5f;
+				}
+
+				if (alignment == center)
+				{
+					//Scan for lines and measure them
+					//Reposition them to the center
+					std::vector<std::pair<u32, u32>> lines;
+					u32 line_begin = 0;
+					u32 ctr = 0;
+
+					for (auto c : text)
+					{
+						switch (c)
+						{
+						case '\r':
+							continue;
+						case '\n':
+							lines.push_back({ line_begin, ctr });
+							line_begin = ctr;
+							continue;
+						default:
+							ctr+=4;
+						}
+					}
+
+					lines.push_back({ line_begin, ctr });
+					const auto max_region_w = std::max<f32>(text_extents_w, w);
```

Comment:
The way "w" is calculated in measure_text() does not take line breaks into account, so max_region_w ends up too large here. 
I think this is what moves everything to the right. In my screenshots you can see that the offset is dependent on the length of the achievement name. Actually text_extents_w should hold the length of the largets line already.

Reply:
Thats the intention. Line breaks are skipped when generating the quads but I might be overlooking something. It does indeed seem to be related to the length of the second line which is strange. I'll run ds2 and get the "This is dark souls trophy" to confirm.

Reply:
Confirmed, length of trophy name causes an offset in X which is wrong.

Reply:
Changing the logic here breaks all other centered text, but the right thing should be to manually set the maximum width of the trophy text during creation. That way the text is centered into a given space (the same way as for message dialogs, e.g. text_display.set_size(1280, 40);)

Reply:
No, label control has autosize method for this very reason; you cannot reliably set the text dimensions as they are affected by the glyphs. Its a bug somewhere else, it certainly should not have an offset in X for the first line unless the second line is longer than the first which it is not in this case. I'm working on it.

Reply:
Ok, found it, its a silly bug in measure_text as you said. Fixing it fixes the centering issue.